controllers/grafana_controller.go

```diff
@@ -111,69 +110,66 @@ func (r *GrafanaReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
 		}
 	}
 
-	for _, stage := range stages {
+	vars := &grafanav1beta1.OperatorReconcileVars{}
+	for _, stage := range getInstallationStages() {
 		log.Info("running stage", "stage", stage)
 
-		nextStatus.Stage = stage
+		grafana.Status.Stage = stage
 		reconciler := r.getReconcilerForStage(stage)
 
 		if reconciler == nil {
 			log.Info("no reconciler known for stage", "stage", stage)
 			continue
 		}
 
-		status, err := reconciler.Reconcile(ctx, grafana, nextStatus, vars, r.Scheme)
+		stageStatus, err := reconciler.Reconcile(ctx, grafana, vars, r.Scheme)
 		if err != nil {
-			log.Error(err, "reconciler error in stage", "stage", stage)
-			nextStatus.LastMessage = err.Error()
-
+			grafana.Status.StageStatus = stageStatus // In progress or failed, both accompanied by Error
+			grafana.Status.LastMessage = err.Error()
 			metrics.GrafanaFailedReconciles.WithLabelValues(grafana.Namespace, grafana.Name, string(stage)).Inc()
-		} else {
-			nextStatus.LastMessage = ""
-		}
-
-		nextStatus.StageStatus = status
 
-		if status != grafanav1beta1.OperatorStageResultSuccess {
-			log.Info("stage in progress", "stage", stage)
-			finished = false
-			break
+			return ctrl.Result{}, fmt.Errorf("reconciler error in stage '%s': %w", stage, err)
 		}
+
+		grafana.Status.StageStatus = grafanav1beta1.OperatorStageResultSuccess
+		grafana.Status.LastMessage = ""
 	}
 
-	if finished {
-		v, err := r.getVersion(ctx, grafana)
-		if err != nil {
-			log.Error(err, "failed to get version from instance")
-		}
-		nextStatus.Version = v
-		log.Info("grafana installation complete")
+	version, err := r.getVersion(ctx, grafana)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to get version from instance: %w", err)
```

Comment:
Previously, a failure to obtain version for an internal grafana would result only in logging the error + removing version from status. Now, you suggest to make the reconciliation fail, but the status will say that it's success due to defer. So, I think the logic here needs to be improved somehow.

Reply:
It's specifically not failing right now, as I also wrote above.
I will add that in a separate PR, I have however improved the handling a bit and added some explanatory comments as to why.